### PR TITLE
[bitnami/mariadb-galera] Fix bootstraping

### PR DESCRIPTION
**Pull request by rafariossaa**
Signed-off-by: Rafael Rios Saavedra <rafael.rios.saavedra@gmail.com>

**Description of the change**

This change fixes the problem of boostraping a new cluster and also the issues related with restarting a already running cluster (or some of its nodes in some cases)

**Benefits**

- Fix issues,
- Simplify bootstraping logic in the chart and in the image.
- Updates the readme for the forcing procedure.

**Possible drawbacks**

Now, there is a new parameter to indicate when you are forcing the boostraping `galera.bootstrap.forceBootstrap` that changes a little bit the previous procedure.

**Applicable issues**

  - fixes #
(This list will be provided)
**Additional information**

**Checklist**

- [X] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/).
- [X] Variables are documented in the values.yaml and added to the README.md using (readme-generator-for-helm)[https://github.com/bitnami-labs/readme-generator-for-helm]
- [X] Title of the PR starts with chart name (e.g. [bitnami/<name_of_the_chart>])
- [X] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/master/CONTRIBUTING.md#sign-your-work)

### CODE REVIEW

File: bitnami/mariadb-galera/templates/statefulset.yaml [hunk 1]
```diff
@@ -98,8 +98,16 @@ spec:
...
+            {{- $fullName := include "common.names.fullname" . }}
+            {{- $releaseName := .Release.Namespace }}
+            {{- $clusterDomain := .Values.clusterDomain }}
+            {{- $initialCluster := list }}
+            {{- range $e, $i := until (int .Values.replicaCount) }}
+            {{- $initialCluster = append $initialCluster (printf "%s-%d.%s-headless.%s.svc.%s" $fullName $i $fullName $releaseName $clusterDomain) }}
+            {{- end }}
             - name: MARIADB_GALERA_CLUSTER_ADDRESS
-              value: "gcomm://{{ template "common.names.fullname" . }}-headless.{{ .Release.Namespace }}.svc.{{ .Values.clusterDomain }}"
+              value: {{ printf "gcomm://%s" (join "," $initialCluster) | quote }}
```
Comment (nicolaerosia) [inline bitnami/mariadb-galera/templates/statefulset.yaml:110]: will this fix ipv6 as well maybe?

File: bitnami/mariadb-galera/pwd.yaml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+apiVersion: v1
+kind: Secret
+metadata:
+  name: mariadb-galera-sec
+type: Opaque
+data:
+  mariadb-root-password: bXlSb290UGFzc3dvcmQK
+  mariadb-password: bXlEYlBhc3N3b3JkCg==
+  mariadb-galera-mariabackup-password: bXlCYWNrdXBQYXNzd29yZAo=
```
Comment (juan131) [inline bitnami/mariadb-galera/pwd.yaml:9]: unintended, right? Please remove it from the PR

File: bitnami/mariadb-galera/templates/statefulset.yaml [hunk 3]
```diff
@@ -8,7 +8,7 @@ spec:
   replicas: {{ .Values.replicaCount }}
   selector:
     matchLabels: {{- include "common.labels.matchLabels" . | nindent 6 }}
-  serviceName: {{ template "common.names.fullname" . }}
+  serviceName: {{ printf "%s-headless" (include "common.names.fullname" .) }}
```
Comment (juan131) [inline bitnami/mariadb-galera/templates/statefulset.yaml:11]: It's always a good practice to ensure we meet the 63 characters limitation on service names (DNS limitation):
```suggestion
  serviceName: {{ printf "%s-headless" (include "common.names.fullname" .) | trunc 63 | trimSuffix "-" }}
```
Please apply the same at https://github.com/bitnami/charts/blob/master/bitnami/mariadb-galera/templates/headless-svc.yaml#L4

File: bitnami/mariadb-galera/templates/statefulset.yaml [hunk 4]
```diff
@@ -98,8 +98,16 @@ spec:
               value: {{ ternary "true" "false" (or .Values.image.debug .Values.diagnosticMode.enabled) | quote }}
             - name: MARIADB_GALERA_CLUSTER_NAME
               value: {{ .Values.galera.name | quote }}
+
```
Comment (juan131) [inline bitnami/mariadb-galera/templates/statefulset.yaml:101]: Remove empty line:
```suggestion
```

File: bitnami/mariadb-galera/values.yaml [hunk 5]
```diff
@@ -239,9 +239,12 @@ galera:
   ## ref: https://github.com/bitnami/bitnami-docker-mariadb-galera#bootstraping
   ##
   bootstrap:
+    ## @param galera.bootstrap.forceBootstrap Option to force the boostraping from the indicated node in `galera.bootstarp.bootstrapFromNode`
+    ##
+    forceBootstrap: false
```
Comment (juan131) [inline bitnami/mariadb-galera/values.yaml:244]: I'm a little bit confused. Could you please explain the difference between `galera.bootstrap.forceBootstrap` and `galera.bootstrap.forceSafeToBootstrap`?

**Review by juan131 — CHANGES_REQUESTED** [review]
Thanks for the PR! Please check my comments.

File: bitnami/mariadb-galera/pwd.yaml (hunk 2 above)
Comment (rafariossaa) [reply bitnami/mariadb-galera/pwd.yaml]: Yes, this was unintended.

File: bitnami/mariadb-galera/templates/statefulset.yaml (hunk 3 above)
Comment (rafariossaa) [reply bitnami/mariadb-galera/templates/statefulset.yaml]: ok

File: bitnami/mariadb-galera/values.yaml (hunk 5 above)
Comment (rafariossaa) [reply bitnami/mariadb-galera/values.yaml:244]: Sure.
- `forceBootstrap` is to force a cluster boostraping from the node indicated in `bootstrapFromNode`. This action is only needed when the cluster is stopped and you need to start it again from a particular node.
- `forceSafeToBootstrap` is to set the flag `safe_to_bootstrap: 1` in the file `/bitnami/mariadb/data/grastate.dat. This is a mariadb file where some info regarding the galera node is stored. That flag indicates if it is safe to bootstrap from the node. This indicates if the node is the most up to date, so it is safe to start this node first and let the other sync from this. If the cluster is abruptly stopped the nodes aren't able to know if they are the latest or not hence all will have '0'. To start it again  you will need to force this to '1' in the node you want to bootstrap from. [Here](https://galeracluster.com/2016/11/introducing-the-safe-to-bootstrap-feature-in-galera-cluster/) you have more info on this topic.

File: bitnami/mariadb-galera/templates/statefulset.yaml (hunk 1 above)
Comment (rafariossaa) [reply bitnami/mariadb-galera/templates/statefulset.yaml:109]: Maybe, would you like to test it and provide feedback once this is merged ?

**Review by juan131 — APPROVED** [review]
LGTM

File: bitnami/mariadb-galera/templates/servicemonitor.yaml [hunk 6]
```diff
@@ -7,7 +7,14 @@ metadata:
   namespace: {{ .Values.metrics.serviceMonitor.namespace }}
   {{- end }}
   labels: {{- include "common.labels.standard" . | nindent 4 }}
+    {{- if .Values.commonLabels }}
+    {{- include "common.tplvalues.render" (dict "value" .Values.commonLabels "context" $) | nindent 4 }}
+    {{- end }}
```
Comment (migruiz4) [inline bitnami/mariadb-galera/templates/servicemonitor.yaml:12]: Please add the value `serviceMonitor.labels`

```suggestion
  labels: {{- include "common.labels.standard" . | nindent 4 }}
    {{- if .Values.commonLabels }}
    {{- include "common.tplvalues.render" (dict "value" .Values.commonLabels "context" $) | nindent 4 }}
    {{- end }}
    {{- if .Values.metrics.serviceMonitor.labels }}
    {{- include "common.tplvalues.render" (dict "value" .Values.metrics.serviceMonitor.labels "context" $) | nindent 4 }}
    {{- end }}
```

File: bitnami/mariadb-galera/templates/svc.yaml [hunk 7]
```diff
@@ -22,11 +31,14 @@ spec:
   {{- end }}
   ports:
     - name: mysql
-      port: {{ .Values.service.port }}
+      port: {{ .Values.service.ports.mysql }}
       targetPort: mysql
-      {{- if and .Values.service.nodePort (or (eq .Values.service.type "NodePort") (eq .Values.service.type "LoadBalancer")) }}
-      nodePort: {{ .Values.service.nodePort }}
+      {{- if and .Values.service.nodePorts.galera (or (eq .Values.service.type "NodePort") (eq .Values.service.type "LoadBalancer")) }}
+      nodePort: {{ .Values.service.nodePorts.galera }}
```
Comment (migruiz4) [inline bitnami/mariadb-galera/templates/svc.yaml:37]: This value should be `.Values.service.nodePorts.mysql`

File: bitnami/mariadb-galera/values.yaml [hunk 8]
```diff
@@ -522,9 +556,25 @@ tolerations: []
...
 ##
+## Galera container port
+## @param containerPorts.mysql mariadb database container port
+## @param containerPorts.galera galera cluster container port
+## @param containerPorts.galera galera IST container port
+## @param containerPorts.galera galera SST container port
+##
+containerPorts:
+  mysql: 3306
+  # TODO: Container image is not aware if the following container's ports
```
Comment (migruiz4) [inline bitnami/mariadb-galera/values.yaml:573]: This sentence is incomplete, and I think it should be something like: 
`# Note: Container image is not aware of changes in the following container's ports`

File: bitnami/mariadb-galera/README.md [hunk 9]
```diff
@@ -532,6 +546,17 @@ $ helm upgrade my-release bitnami/mariadb-galera \
...
 | Note: you need to substitute the placeholders _[ROOT_PASSWORD]_, _[MARIADB_PASSWORD]_ and _[MARIABACKUP_PASSWORD]_ with the values obtained from instructions in the installation notes.
 
+### To 7.0.0
+
+This major release renames several values in this chart and adds missing features, in order to be inline with the rest of assets in the Bitnami charts repository. Also, this release address issues related to cluster initialization and node's restarts.
+
+Affected values:
+
+- `service.port` renamed as `service.ports.mysql`.
+- `service.nodePort` renamed as `service.nodePorts`.
```
Comment (migruiz4) [inline bitnami/mariadb-galera/README.md:556]:
```suggestion
- `service.nodePort` renamed as `service.nodePorts.mysql`.
```

File: bitnami/mariadb-galera/README.md [hunk 10]
```diff
@@ -532,6 +546,17 @@ $ helm upgrade my-release bitnami/mariadb-galera \
...
+- `securityContext` renamed as `podSecurityContext`.
+- `extraInitContainers` renamed as `initContainers`.
+
```
Comment (migruiz4) [inline bitnami/mariadb-galera/README.md:559]: Once done, please add 

```
- `prometheusRule.selector` renamed as `prometheusRule.additionalLabels`
```

**Review by migruiz4 — CHANGES_REQUESTED** [review]
Hi @rafariossaa,

Please check my comments below.

In addition, I'm missing `commonAnnotations` in most of the templates, and don't forget to update the README.md.

```
  {{- if .Values.commonAnnotations }}
  annotations: {{- include "common.tplvalues.render" (dict "value" .Values.commonAnnotations "context" $) | nindent 4 }}
  {{- end }}
```

File: bitnami/mariadb-galera/README.md [hunk 11]
```diff
@@ -532,6 +572,18 @@ $ helm upgrade my-release bitnami/mariadb-galera \
...
+This major release renames several values in this chart and adds missing features, in order to be inline with the rest of assets in the Bitnami charts repository. Also, this release address issues related to cluster initialization and node's restarts.
+
+Affected values:
+
+- `service.port` renamed as `service.ports.mysql`.
+- `service.nodePort` renamed as `service.nodePorts.mysql`.
+- `securityContext` renamed as `podSecurityContext`.
+- `extraInitContainers` renamed as `initContainers`.
+- `prometheusRule.selector` renamed as `prometheusRule.additionalLabels`
+
```
Comment (juan131) [inline bitnami/mariadb-galera/README.md:586]: Am I supposed to run any special steps to upgrade from `6.x.x`? If so, please document them.

File: bitnami/mariadb-galera/values.yaml [hunk 12]
```diff
@@ -142,15 +157,22 @@ service:
...
 ##
 serviceAccount:
-  ## @param serviceAccount.create Specify whether a ServiceAccount should be created
-  ##
   create: false
-  ## @param serviceAccount.name The name of the ServiceAccount to create
-  ## If not set and create is true, a name is generated using the common.names.fullname template
-  ##
   name: ""
+  automountServiceAccountToken: true
```
Comment (juan131) [inline bitnami/mariadb-galera/values.yaml:168]: Does MariaDB Galera makes uses of the K8s API by default? If not, this should be set to "false" by default.

File: bitnami/mariadb-galera/README.md (hunk 11 above)
Comment (rafariossaa) [reply bitnami/mariadb-galera/README.md:586]: No special steps are needed.

**Pull request merged**